### PR TITLE
ArchiveFactory fails to close ZipFile and retains a Windows file lock

### PR DESCRIPTION
**Pull request by AndyGee**

### CODE REVIEW

**Comment by ALRubinger** [comment]
```
Tests in error: 
  ShrinkWrapTestCase.shouldBeAbleToimportZipFileViaArchiveFactory:362 » Runtime ...
  ShrinkWrapTestCase.shouldBeAbleToimportZipFileViaShrinkWrap:343 » Runtime Coul...
  ZipExporterTestCase.testExportImportExport:190 » ArchiveExport Failed to write...
```

File: api/src/main/java/org/jboss/shrinkwrap/api/ArchiveFactory.java [hunk 1]
```diff
@@ -184,19 +184,27 @@
...
+            if(null != zipFile){
+                try {
+                    zipFile.close();
+                } catch (final IOException e) {
+                    //no-op
+                }
+            }
         }
 
-        // Import
```
Comment (ALRubinger) [inline api/src/main/java/org/jboss/shrinkwrap/api/ArchiveFactory.java]: We remove the return statement?

Comment (AndyGee) [inline api/src/main/java/org/jboss/shrinkwrap/api/ArchiveFactory.java]: Moved into try block

File: api/src/main/java/org/jboss/shrinkwrap/api/ArchiveFactory.java [hunk 2]
```diff
@@ -184,19 +184,27 @@
         }
 
         // Construct ZipFile
-        final ZipFile zipFile;
+        ZipFile zipFile = null;
         try {
             zipFile = new ZipFile(archiveFile);
+            // Import
+            return ShrinkWrap.create(type, archiveFile.getName()).as(ZipImporter.class).importFrom(zipFile).as(type);
```
Comment (AndyGee) [inline api/src/main/java/org/jboss/shrinkwrap/api/ArchiveFactory.java]: Return here, and close ZipFile in finally block

**Comment by AndyGee** [comment]
OK, so I can see that ZipFileEntryAsset requires the entry not to be closed, but there is nothing to ensure that it ever gets closed.

**Comment by AndyGee** [comment]
OK, this does it. ZipFileEntryAsset now gets a file and creates a new ZipFile that gets closed in the InputStreamWrapper.

File: impl-base/src/main/java/org/jboss/shrinkwrap/impl/base/importer/zip/ZipImporterImpl.java [hunk 3]
```diff
@@ -214,10 +214,16 @@ private ZipImporter importFrom(final ZipFile file, Filter<ArchivePath> filter) t
                     continue;
                 }
 
-                archive.add(new ZipFileEntryAsset(file, entry), new BasicPath(entryName));
+                archive.add(new ZipFileEntryAsset(new File(file.getName()), entry), new BasicPath(entryName));
```
Comment (gastaldi) [inline impl-base/src/main/java/org/jboss/shrinkwrap/impl/base/importer/zip/ZipImporterImpl.java]: Why not just `new ZipFileEntryAsset(file,entry)`?

File: api/src/main/java/org/jboss/shrinkwrap/api/ArchiveFactory.java [hunk 4]
```diff
@@ -183,20 +183,15 @@
                 + archiveFile.getAbsolutePath());
         }
 
-        // Construct ZipFile
-        final ZipFile zipFile;
         try {
-            zipFile = new ZipFile(archiveFile);
+            // Import
+            return ShrinkWrap.create(type, archiveFile.getName()).as(ZipImporter.class).importFrom(new ZipFile(archiveFile)).as(type);
```
Comment (gastaldi) [inline api/src/main/java/org/jboss/shrinkwrap/api/ArchiveFactory.java]: What changed here? The try-catch? Should we catch ZIP exceptions here?

**Comment by ALRubinger** [comment]
eccccd10005dbc7e63d780e5f40b437963b92d64

File: impl-base/src/main/java/org/jboss/shrinkwrap/impl/base/importer/zip/ZipImporterImpl.java (hunk 3 above)
Comment (AndyGee) [inline impl-base/src/main/java/org/jboss/shrinkwrap/impl/base/importer/zip/ZipImporterImpl.java]: The ZipFileEntryAsset now just gets a simple java.io.File and ZipImporterImpl closes the original ZipFile (which was used for iteration and filtering) - This releases the Windows file handle.

A call to ZipFileEntryAsset#openStream() now uses the java.io.File to create a new ZipFile and stream, which is wrapped. Closing this wrapped stream now also closes the ZipFile, again releasing the Windows handle. So each call to ZipFileEntryAsset#openStream() repeats this, each time ensuring the Windows handle is released safely.

File: api/src/main/java/org/jboss/shrinkwrap/api/ArchiveFactory.java (hunk 4 above)
Comment (AndyGee) [inline api/src/main/java/org/jboss/shrinkwrap/api/ArchiveFactory.java]: This was just inlining the new ZipFile as the variable is unused, and cannot be closed at this point..

**Comment by AndyGee** [comment]
Thanks for the merge @ALRubinger.

**Comment by ALRubinger** [comment]
Thank you for the report and fix, @AndyGee !

**Pull request closed without merging**